In [1]:
import torch
import torch.nn as nn
device='cpu'
#device='cuda'

In [2]:
class LinNet(nn.Module):
    def __init__(self):
        super(LinNet, self).__init__()
        # Define the model.
        self.layer1 = nn.Sequential(nn.Linear(in_features=2, out_features=2, bias=True))
        # Generate a fully connected linear neural network model, 1 layer, bias, linear activation function
        # returns: Trainable object
        #self.act = nn.LeakyReLU() #non-linear activation function
        #self.act = nn.ReLU() #non-linear activation function
    
    def forward(self, x):
        out = self.layer1(x)
        #out = self.act(out) #comment out if not desired
        return out

Data_preparation

In [3]:
#input tensor, type torch tensor:
#Indices: batch, sample, features or signal dimension. Here: 1 batch, 3 samples, signal dimension 2:

#Training set:
X=torch.tensor([[1., 2.], [2., 1.],[1., 1.]]).view(1,3,2) #adding the first dimension for the batch
print("X.shape", X.shape)

#Target:
Y=torch.tensor([[1., 0.], [0., 1.],[0., 0.]]).view(1,3,2)
print("Y.shape", Y.shape)

#Validation set, to test generalization:
Xval=torch.tensor([[0.5, 1.0], [1., 0.5],[0.5, 0.5]]).view(1,3,2)
#Validation Target:
Yval=torch.tensor([[1., 0.], [0., 1.],[0., 0.]]).view(1,3,2)

X.shape torch.Size([1, 3, 2])
Y.shape torch.Size([1, 3, 2])


In [5]:
#create network object:
model = LinNet().to(device)
loss_fn = nn.MSELoss()
print("Define loss function:", loss_fn)
#learning_rate = 1e-4
#optimizer = torch.optim.Adam(model.parameters())
optimizer = torch.optim.SGD(model.parameters(),lr=0.1)
print("Define optimizer:", optimizer)

Define loss function: MSELoss()
Define optimizer: SGD (
Parameter Group 0
    dampening: 0
    lr: 0.1
    momentum: 0
    nesterov: False
    weight_decay: 0
)


In [6]:
for epoch in range(10000):
    Ypred=model(X) #the model produces prediction output
    loss=loss_fn(Ypred, Y) #prediction and target compared by loss
    if epoch%1000==0:
        print(epoch, loss.item()) #print current loss value
    optimizer.zero_grad() #optimizer sets previous gradients to zero
    loss.backward() #optimizer computes new gradients
    optimizer.step() #optimizer updates weights

0 1.3473855257034302
1000 0.00042512454092502594
2000 3.525961801642552e-06
3000 2.926994291385654e-08
4000 2.440122270908063e-10
5000 4.4426684553400264e-12
6000 4.4426684553400264e-12
7000 4.4426684553400264e-12
8000 4.4426684553400264e-12
9000 4.4426684553400264e-12


In [11]:
Ypred=model(X) # Make Predictions based on the obtained weights
loss=loss_fn(Ypred, Y)
print("Loss on trainig set:", loss)
Yvalpred=model(Xval) # Make Predictions based on the obtained weights
loss=loss_fn(Yvalpred, Yval)
print("Loss on validation set:", loss)
weights = model.state_dict() #read obtained weights

Loss on trainig set: tensor(4.4427e-12, grad_fn=<MseLossBackward0>)
Loss on validation set: tensor(0.5000, grad_fn=<MseLossBackward0>)


The loss values: Only **4.6493e-12** on the **training se**t, but **0.5000** on the **validation** set Hence we have a **bad generalization**.